src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs

```diff
@@ -171,28 +173,40 @@ public IReadOnlyList<TagSpan<IntraTextAdornmentTag>> GetTags(NormalizedSnapshotS
                 var classify = document != null && _taggerProvider.EditorOptionsService.GlobalOptions.GetOption(InlineHintsViewOptionsStorage.ColorHints, document.Project.Language);
 
                 var selectedSpans = new List<TagSpan<IntraTextAdornmentTag>>();
-                for (var i = 0; i < _cache.Count; i++)
+                var itemsToUpdate = new List<(IMappingTagSpan<InlineHintDataTag> mappingTagSpan, TagSpan<IntraTextAdornmentTag> hintTagSpan)>();
+
+                foreach (var oldCacheItem in _cache)
                 {
-                    var tagSpans = _cache[i].mappingTagSpan.Span.GetSpans(snapshot);
+                    var mappingTagSpan = oldCacheItem.mappingTagSpan;
+                    var cacheTagSpan = oldCacheItem.tagSpan;
+                    var tagSpans = mappingTagSpan.Span.GetSpans(snapshot);
                     if (tagSpans.Count == 1)
                     {
                         var tagSpan = tagSpans[0];
                         if (spans.IntersectsWith(tagSpan))
                         {
-                            if (_cache[i].tagSpan is not { } hintTagSpan)
+                            if (cacheTagSpan is not { } hintTagSpan)
                             {
                                 var hintUITag = InlineHintsTag.Create(
-                                        _cache[i].mappingTagSpan.Tag.Hint, Format, _textView, tagSpan, _taggerProvider, _formatMap, classify);
+                                        mappingTagSpan.Tag.Hint, Format, _textView, tagSpan, _taggerProvider, _formatMap, classify);
 
                                 hintTagSpan = new TagSpan<IntraTextAdornmentTag>(tagSpan, hintUITag);
-                                _cache[i] = (_cache[i].mappingTagSpan, hintTagSpan);
+                                itemsToUpdate.Add((mappingTagSpan, hintTagSpan));
                             }
 
                             selectedSpans.Add(hintTagSpan);
                         }
                     }
                 }
 
+                // Go through all the items that need to be updated and update the cache since it can't be done
+                // while iterating over it
+                foreach (var (mappingTagSpan, hintTagSpan) in itemsToUpdate)
+                {
+                    _cache.Remove((mappingTagSpan, null));
+                    _cache.Add((mappingTagSpan, hintTagSpan));
+                }
```

Comment:
what is going on here?  why are we adding/removing anythnig from the cache?  I don't get it as the cache is only used when teh snapshot stays the same (otherwise it is cleared).  so what purpose does this serve?